shared/typetracking/codeql/typetracking/TypeTracking.qll

```diff
@@ -1,69 +1,240 @@
+    WithContent(ContentFilter filter) { withContentStep(_, _, filter) } or
+    WithoutContent(ContentFilter filter) { withoutContentStep(_, _, filter) } or
     JumpStep()
 
-  cached
-  newtype TTypeTracker =
-    MkTypeTracker(Boolean hasCall, OptionalTypeTrackerContent content) {
-      content = noContent()
+  /**
+   * A description of a step on an inter-procedural data flow path.
+   */
+  private class StepSummary extends TStepSummary {
+    /** Gets a textual representation of this step summary. */
+    string toString() {
+      this instanceof LevelStep and result = "level"
+      or
+      this instanceof CallStep and result = "call"
+      or
+      this instanceof ReturnStep and result = "return"
+      or
+      exists(Content content | this = StoreStep(content) | result = "store " + content)
+      or
+      exists(Content content | this = LoadStep(content) | result = "load " + content)
+      or
+      exists(Content load, Content store |
+        this = LoadStoreStep(load, store) and
+        result = "load-store " + load + " -> " + store
+      )
+      or
+      this instanceof JumpStep and result = "jump"
+    }
+  }
+
+  private newtype TTypeTracker =
+    MkTypeTracker(Boolean hasCall, ContentOption content) {
+      content.isNone()
       or
       // Restrict `content` to those that might eventually match a load.
       // We can't rely on `basicStoreStep` since `startInContent` might be used with
       // a content that has no corresponding store.
-      exists(TypeTrackerContent loadContents |
+      exists(Content loadContents |
         (
-          basicLoadStep(_, _, loadContents)
+          loadStep(_, _, loadContents)
           or
-          basicLoadStoreStep(_, _, loadContents, _)
+          loadStoreStep(_, _, loadContents, _)
         ) and
-        compatibleContents(content, loadContents)
+        compatibleContents(content.asSome(), loadContents)
       )
     }
 
-  cached
-  newtype TTypeBackTracker =
-    MkTypeBackTracker(Boolean hasReturn, OptionalTypeTrackerContent content) {
-      content = noContent()
+  private newtype TTypeBackTracker =
+    MkTypeBackTracker(Boolean hasReturn, ContentOption content) {
+      content.isNone()
       or
       // As in MkTypeTracker, restrict `content` to those that might eventually match a store.
-      exists(TypeTrackerContent storeContent |
+      exists(Content storeContent |
         (
-          basicStoreStep(_, _, storeContent)
+          storeStep(_, _, storeContent)
           or
-          basicLoadStoreStep(_, _, _, storeContent)
+          loadStoreStep(_, _, _, storeContent)
         ) and
-        compatibleContents(storeContent, content)
+        compatibleContents(storeContent, content.asSome())
       )
     }
 
   pragma[nomagic]
   private TypeTracker noContentTypeTracker(boolean hasCall) {
-    result = MkTypeTracker(hasCall, noContent())
+    result = MkTypeTracker(hasCall, any(ContentOption::None c))
```

Comment:
Is there a `ContentOption::none()` injector that can be used instead?

Reply:
No, `none()` is a builtin predicate in QL, so we' can't have it in `Option`.

Reply:
Ah, yes 🤦